### PR TITLE
[AutoParallel] add grad_clip pass

### PR DESCRIPTION
**Pull request by zhaoyinglia**
### PR types

New features
### PR changes

Others
### Describe

- add grad_clip pass for auto_parallel. 
- each rank computes its own norm value, then gets global_norm by allreduce_sum only once. 
image

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/parallelizer_v2.py [hunk 1]
```diff
@@ -217,7 +217,19 @@ def _apply_post_optimization(self, main_program, startup_program, rank,
             auto_parallel_sharding_pass.apply([main_program], [startup_program],
                                               self._pass_context)
 
-        # recompute is then train-only optimization
+        # GradClip is train-only optimization
+        grad_clip = self._dist_context._lr_optimizer._grad_clip
```
Comment (JZ-LIANG) [inline python/paddle/distributed/auto_parallel/parallelizer_v2.py]: better to implement this logic by the _check_sell() function in PassBase, while dist_context is already a argument to PassBase

File: python/paddle/distributed/passes/auto_parallel_grad_clip.py [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+                                'dtype': input_var.dtype,
+                                'value': 0,
+                                'force_cpu': False,
+                                OP_ROLE_KEY: OpRole.Forward
+                            })
+                        fill_constant_op._set_attr('op_namescope',
+                                                   "/gradient_clip_pass")
+                        offset += 1
+
+                    allreduce_op = block._insert_op(
```
Comment (JZ-LIANG) [inline python/paddle/distributed/passes/auto_parallel_grad_clip.py:327]: dist attr for new op and tensor

File: python/paddle/distributed/passes/auto_parallel_grad_clip.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+                            idx,
+                            type='fill_constant',
+                            inputs={},
+                            outputs={'Out': [input_var]},
+                            attrs={
+                                'shape': [1],
+                                'dtype': input_var.dtype,
+                                'value': 0,
+                                'force_cpu': False,
+                                OP_ROLE_KEY: OpRole.Forward
```
Comment (JZ-LIANG) [inline python/paddle/distributed/passes/auto_parallel_grad_clip.py]: OpRole.Forward ?

Comment (zhaoyinglia) [reply python/paddle/distributed/passes/auto_parallel_grad_clip.py]: fixed.

File: python/paddle/distributed/passes/auto_parallel_grad_clip.py (hunk 2 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/passes/auto_parallel_grad_clip.py:327]: done.

File: python/paddle/distributed/auto_parallel/parallelizer_v2.py (hunk 1 above)
Comment (zhaoyinglia) [reply python/paddle/distributed/auto_parallel/parallelizer_v2.py]: done.

**Review by XieYunshen — APPROVED** [review]
LGTM for
` set_tests_properties(test_grad_clip PROPERTIES LABELS "RUN_TYPE=EXCLUSIVE"`

**Review by JZ-LIANG — APPROVED** [review]
LGTM

**Pull request merged**